### PR TITLE
Marshal mapping

### PR DESCRIPTION
**Pull request by olejnjak**
Introduce `mapResponseMarshal` extension for `SignalProducer`.

### CODE REVIEW

File: ACKReactiveExtensions/Marshal/MarshalMapping.swift [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+     */
+    public func mapResponseMarshal<Model>(for key: KeyType? = nil) -> Signal<[Model], Error> where Model: Unmarshaling {
+        return attemptMap { json in
+            Result {
+                if let key = key, let marshaledJSON = json as? MarshaledObject {
+                    return try marshaledJSON.value(for: key)
+                }
+                else if let marshaledArray = json as? [MarshaledObject] {
+                    let dummyKey = "dummyKey"
+                    return try [dummyKey: marshaledArray].value(for: dummyKey)
```
Comment (tkohout) [inline ACKReactiveExtensions/Marshal/MarshalMapping.swift]:
```
try marshaledArray.map (Model.init)
```

File: ACKReactiveExtensions/Marshal/MarshalMapping.swift [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    /**
+     * Create error containing passed `MarshalError`
+     *
+     * - parameter marshalError: `MarshalError` which should be wrapped
+     */
+    static func createMarshalError(_ marshalError: MarshalError) -> Self
+}
+
+extension MarshalError: MarshalErrorCreatable {
+    public static func createMarshalError(_ marshalError: MarshalError) -> MarshalError {
```
Comment (tkohout) [inline ACKReactiveExtensions/Marshal/MarshalMapping.swift:27]: Shouldn't this be `_ marshalError: NSError`?

File: ACKReactiveExtensions/Marshal/MarshalMapping.swift [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+     *
+     * - parameter key: If your objects are contained within dictionary pass the key here
+     */
+    public func mapResponseMarshal<Model>(for key: KeyType? = nil) -> Signal<Model, Error> where Model: Unmarshaling {
+        return attemptMap { json in
+            Result {
+                guard let marshaledJSON = json as? MarshaledObject
+                    else {
+                        assertionFailure("json isn't any of the known MarshaledObject types (Dictionary or Array)")
+                        throw NSError(domain: "", code: 0, userInfo: nil)
```
Comment (tkohout) [inline ACKReactiveExtensions/Marshal/MarshalMapping.swift]: Could we get rid of this and throw some better error?

File: ACKReactiveExtensions/Marshal/MarshalMapping.swift [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    }
+}
+
+extension SignalProducerProtocol where Value == Any, Error: MarshalErrorCreatable {
+    /**
+     * Map value as `Unmarshaling` object
+     *
+     * - parameter key: If your objects are contained within dictionary pass the key here
+     */
+    public func mapResponseMarshal<Model>(for key: KeyType? = nil) -> SignalProducer<Model, Error> where Model: Unmarshaling {
```
Comment (tkohout) [inline ACKReactiveExtensions/Marshal/MarshalMapping.swift]: I would change it to `forKey key: KeyType? = nil)`. It's not clear what for: means.

File: ACKReactiveExtensions/Marshal/MarshalMapping.swift [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    
+    /**
+     * Map value as `ValueType`
+     *
+     * - parameter key: If your objects are contained within dictionary pass the key here
+     */
+    public func mapResponseMarshal<Model>(for key: KeyType) -> SignalProducer<Model, Error> where Model: ValueType {
+        return lift { $0.mapResponseMarshal(for: key) }
+    }
+}
```
Comment (tkohout) [inline ACKReactiveExtensions/Marshal/MarshalMapping.swift]: It would be also nice to have some test to call the public API

**Comment by olejnjak** [comment]
Conflicts resolved

File: ACKReactiveExtensions/Marshal/MarshalMapping.swift (hunk 2 above)
Comment (olejnjak) [reply ACKReactiveExtensions/Marshal/MarshalMapping.swift:27]: Don't see any reason why? As far as Marshal is throwing `MarshalError`s.

File: ACKReactiveExtensions/Marshal/MarshalMapping.swift (hunk 5 above)
Comment (olejnjak) [reply ACKReactiveExtensions/Marshal/MarshalMapping.swift]: You mean something like [this](https://github.com/AckeeCZ/ACKReactiveExtensions/blob/8bb0cc8dca4eb357b302234e78d01e6361ae2077/Example/Tests/MarshalMappingTests.swift)?

File: ACKReactiveExtensions/Marshal/MarshalMapping.swift (hunk 2 above)
Comment (olejnjak) [reply ACKReactiveExtensions/Marshal/MarshalMapping.swift:27]: Aaaaa I see...

Comment (olejnjak) [reply ACKReactiveExtensions/Marshal/MarshalMapping.swift:27]: Now when all errors thrown by the extension are really `MarshalError`s this is correct.

**Pull request merged**